File: recursive_sample.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
)

func globPaths(targetPath string, maxDepth int, suffix string, depth int) []string {
	var matchedPaths []string

	if depth == maxDepth {
		fmt.Printf("Reached maximum depth of %d at path %s\n", maxDepth, targetPath)
	} else {
		fileInfo, err := os.Stat(targetPath)
		if err != nil {
			panic(err)
		}

		if fileInfo.IsDir() {
			depth++

			fileInfos, err := ioutil.ReadDir(targetPath)
			if err != nil {
				panic(err)
			}

			for _, fileInfo := range fileInfos {
				subPath := filepath.Join(targetPath, fileInfo.Name())
				subPathMatchedPaths := globPaths(subPath, maxDepth, suffix, depth)
				matchedPaths = append(matchedPaths, subPathMatchedPaths...)
			}
		} else if fileInfo.Mode().IsRegular() {
			if filepath.Ext(targetPath) == suffix {
				matchedPaths = append(matchedPaths, targetPath)
			}
		}
	}

	return matchedPaths
}

func main() {
	result := globPaths(".", 2, ".json", 0)
	for _, item := range result {
		fmt.Println(filepath.Join(filepath.Dir(item), fmt.Sprintf("%s.json5", filepath.Base(item[:len(item)-len(filepath.Ext(item))]))))
	}
}

File: recursive_sample.py
import os
from typing import List

def glob_paths(target_path: str, max_depth: int, *, suffix: str, depth: int = 0) -> List[str]:
    matched_paths: List[str] = []
    if depth == max_depth:
        pass
        print(f'引数からのディレクトリの深さ{depth}/{max_depth}階層、ファイルPATH:{target_path}')
    elif os.path.isdir(target_path):
        depth += 1
        for list_item in os.listdir(target_path):
            sub_path = os.path.join(target_path, list_item)
            subpath_matched_paths: List[str] = glob_paths(sub_path, max_depth, suffix=suffix, depth=depth)
            matched_paths.extend(subpath_matched_paths)
    elif os.path.isfile(target_path):
        if target_path.endswith(suffix):
            matched_paths.append(target_path)
    return matched_paths

def main():
    result = glob_paths(target_path=os.getcwd(), max_depth=2, suffix=".json")
    for item in result:
        print(os.path.splitext(item)[0] + ".json5")
        os.path
main()
